csl/core/engine.lua

```diff
@@ -0,0 +1,1010 @@
+   local and_opt = name_node.options["and"] or "text"
+   local and_word = and_opt == "symbol" and "&" or self.locale:term("and") -- text by default
+   local name_delimiter = name_node.options.delimiter or inherited_opts["names-delimiter"]
+   -- local delimiter_precedes_et_al = name_node.options["delimiter-precedes-et-al"] -- TODO NOT IMPLEMENTED
+
+   local resolved = {
+      variable = SU.required(name_node.options, "variable", "CSL names"),
+      et_al_min = et_al_min,
+      et_al_use_first = et_al_use_first,
+      and_word = and_word,
+      name_delimiter = name_delimiter,
+      is_label_first = is_label_first,
+      label_opts = label_opts,
+      et_al_opts = et_al_opts,
+      name_node = name_node,
+      names_delimiter = options.delimiter or inherited_opts["names-delimiter"],
+   }
+   resolved = pl.tablex.union(options, resolved)
+
+   return self:_names_with_resolved_opts(resolved, substitute, entry)
+end
+
+function CslEngine:_label (options, content, entry)
+   local variable = SU.required(options, "variable", "CSL label")
+   local value = entry[variable]
+   self:_addGroupVariable(variable, value)
+   if variable == "locator" then
+      variable = value and value.label
+      value = value and value.value
+   end
+   if value then
+      local plural = options.plural
+      if plural == "always" then
+         plural = true
+      elseif plural == "never" then
+         plural = false
+      else -- "contextual" by default
+         if variable == "number-of-pages" or variable == "number-of-volumes" then
+            local  v = tonumber(value)
+            plural = v and v > 1 or false
+         else
+            if type(value) == "table" then
+               plural = #value > 1
+            else
+               local _, count = string.gsub(tostring(value), "%d+", "") -- naive count of numbers
+               plural = count > 1
+            end
+         end
+      end
+      value = self.locale:term(variable, options.form or "long", plural)
+      value = self:_render_stripPeriods(value, options)
+      value = self:_render_textCase(value, options)
+      value = self:_render_formatting(value, options)
+      value = self:_render_affixes(value, options)
+      return value
+   end
+   return value
+end
+
+function CslEngine:_group (options, content, entry)
+   self:_enterGroup()
+
+   local t = self:_render_children(content, entry, { delimiter = options.delimiter })
+   t = self:_render_formatting(t, options)
+   t = self:_render_affixes(t, options)
+   t = self:_render_display(t, options)
+
+   t = self:_leaveGroup(t) -- Takes care of group suppression
+   return t
+end
+
+function CslEngine:_if (options, content, entry)
+   local match = options.match or "all"
+   local conds = {}
+   if options.variable then
+      local vars = pl.stringx.split(options.variable, " ")
+      local cond = false
+      for _, var in ipairs(vars) do
+         if entry[var] then
+            cond = true
+            break
```

Comment:
Testing with several style files and a variety of inputs, my reading of the spec here is wrong. here and below... I understood it as "evaluate each condition independently" and then "apply the match attribute" to each condition", but it should be applied to each elementary item (e.g. here, each variable).

E.g. seen with chicago-fullnote-bibliography (not with chicago-author-date, but it might just mean my inputs are not representative enough), on `<if variable="original-date issued" match="all">`

Reply:
Note that the items in `type` attribute are always applied with `any` logic (i.e. "or"), which is not clearly stated either. For example, `<if type="chapter paper-conference" variable="container-title" match="all">` means `(type="chapter" or type="paper-conference") and variable="container-title"`.

I also recommend the <https://github.com/citation-style-language/test-suite> repository for testing.

Reply:
Thanks a lot for the explanation, @zepinglee ! It's really appreciated.

(I think the existing spec could be improved to mention such subtleties and be much more clear at some places -- But please take it as a bare statement, not a criticism: I know it's real hard to write good documentations and specifications on such topics.)

N.B. As for the test suite, yes indeed, I am aware of it, and at some point it would need to be addressed for complete conformance. But I'll be real honest here (and I'll clarify it in the PR description depending on the point I eventually reach), I am first of all thriving at getting enough features for processing my next book, incl. BibTeX-based bibliography of ~1500 entries. The "current" experimental bibliography solution in SILE completely fails at it, hence my attempt at rewriting it with CSL in mind. Clearly it's an iterative process with that goal in mind. (and I've only processed ~100 entries so far, with 2-3 style files and the two target locales only, noting pn the way where it's still wrong -- This is also my way at understanding progressively the expectations from the CSL spec, regardless of its test suite).

Reply:
Done